src/driver/iour/mod.rs

```diff
@@ -79,32 +76,57 @@ impl Driver {
                 },
             }?;
 
-            for entry in self.inner.completion() {
-                let entry = create_entry(entry);
-                if entry.user_data() == u64::MAX as _ {
-                    // This is a cancel operation.
-                    continue;
-                }
-                if let Err(e) = &entry.result {
-                    if e.raw_os_error() == Some(libc::ECANCELED) {
-                        // This operation is cancelled.
-                        continue;
-                    }
+            let completed_entries = self.inner.completion().filter_map(|entry| {
+                const SYSCALL_ECANCELED: i32 = -libc::ECANCELED;
+                match (entry.user_data(), entry.result()) {
+                    // Cancel or cancelled operation.
+                    (Self::CANCEL, _) | (_, SYSCALL_ECANCELED) => None,
+                    (..) => Some(create_entry(entry)),
                 }
-                self.cqueue.push_back(entry);
-            }
+            });
+            self.cqueue.extend(entries.extend(completed_entries));
 
             if self.squeue.is_empty() && self.inner.submission().is_empty() {
                 break;
             }
         }
-        Ok(())
+        Ok(entries.entries_len())
+    }
+
+    fn flush_submissions(&mut self) {
+        let mut inner_squeue = self.inner.submission();
+        if inner_squeue.is_full() {
+            // can't flush
+            return;
+        }
+        let remain_space = inner_squeue.capacity() - inner_squeue.len();
+        if self.squeue.len() <= remain_space {
+            // inner queue has enough space for all entries
+            // use batched submission optimization
+            let (s1, s2) = self.squeue.as_slices();
+            unsafe {
+                inner_squeue
+                    .push_multiple(s1)
```

Comment:
We push to the back of the VecDeque and pop from the front.

When it's wrapped the oldest submissions are at the end of the inner buffer. They should be pushed first to preserve the submission order.

Reply:
https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.as_slices

The doc says that the elements are in order.

Reply:
True - https://doc.rust-lang.org/src/alloc/collections/vec_deque/mod.rs.html#1271

Sorry, I didn't check this.